BIP 174 PSBT Serializations and RPCs

achow101 opened the pull request:
This Pull Request fully implements the [updated](https://github.com/bitcoin/bips/pull/694) BIP 174 specification. It is based upon #13425 which implements the majority of the signing logic.

BIP 174 specifies a binary transaction format which contains the information necessary for a signer to produce signatures for the transaction and holds the signatures for an input while the input does not have a complete set of signatures.

This PR contains structs for PSBT, serialization, and deserialzation code. Some changes to `SignatureData` have been made to support detection of UTXO type and storing public keys.

***

Many RPCs have been added to handle PSBTs.

`walletprocesspsbt` takes a PSBT format transaction, updates the PSBT with any inputs related to this wallet, signs, and finalizes the transaction. There is also an option to not sign and just update.

`walletcreatefundedpsbt` creates a PSBT from user provided data in the same form as createrawtransaction. It also funds the transaction and takes an options argument in the same form as `fundrawtransaction`. The resulting PSBT is blank with no input or output data filled in. It is analogous to a combination of `createrawtransaction` and `fundrawtransaction`

`decodepsbt` takes a PSBT and decodes it to JSON. It is analogous to `decoderawtransaction`

`combinepsbt` takes multiple PSBTs for the same tx and combines them. It is analogous to `combinerawtransaction`

`finalizepsbt` takes a PSBT and finalizes the inputs. If all inputs are final, it extracts the network serialized transaction and returns that instead of a PSBT unless instructed otherwise.

`createpsbt` is like `createrawtransaction` but for PSBTs instead of raw transactions.

`convertpsbt` takes a network serialized transaction and converts it into a psbt. The resulting psbt will lose all signature data and an explicit flag must be set to allow transactions with signature data to be converted.

***

This supersedes #12136

achow101 committed:
Implement PSBT Structures and un/serialization methods per BIP 174

achow101 committed:
Add pubkeys and whether input was witness to SignatureData

Stores pubkeys in SignatureData and retrieves them when using GetPubKey().

Stores whether the signatures in a SignatureData are for a witness input.

achow101 committed:
Methods for interacting with PSBT structs

Added methods which move data to/from SignaturData objects to
PSBTInput and PSBTOutput objects.

Added sanity checks for PSBTs as a whole which are done immediately
after deserialization.

Added Merge methods to merge a PSBT into another one.

achow101 committed:
Refactor transaction creation and transaction funding logic

In preparation for more create transaction and fund transcation RPCs,
refactor the transaction creation and funding logic into separate
functions.

achow101 committed:
Tests for PSBT

Added functional tests for PSBT that test the RPCs. Also added all
of the BIP 174 test vectors (except for the updater tests) in the
functional tests.

Added a Unit test for the BIP 174 updater test vector.

promag commented on src/script/sign.cpp:
```diff
@@ -169,15 +171,106 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+        unsigned int num_pubkeys = solutions.size()-2;
+        unsigned int last_success_key = 0;
+        for (const valtype& sig : stack.script) {
+            for (unsigned int i = last_success_key; i < num_pubkeys; i++) {
```
Commit "Make SignatureData able to store signatures and scripts"

nit, `++i`

achow101 replied:
Done

promag commented on src/script/sign.h:
```diff
@@ -81,4 +91,15 @@ void UpdateInput(CTxIn& input, const SignatureData& data);
  * Solvability is unrelated to whether we consider this output to be ours. */
 bool IsSolvable(const SigningProvider& provider, const CScript& script);
 
+class SignatureExtractorChecker : public BaseSignatureChecker
```
Commit  "Make SignatureData able to store signatures and scripts"

`class SignatureExtractorChecker final`?

achow101 replied:
Done

promag commented on src/script/sign.cpp:
```diff
@@ -169,15 +171,106 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+{
+    if (checker->CheckSig(scriptSig, vchPubKey, scriptCode, sigversion)) {
+        CPubKey pubkey(vchPubKey);
+        sigdata->signatures.emplace(pubkey.GetID(), SigPair(pubkey, scriptSig));
```
Commit "Make SignatureData able to store signatures and scripts"

This is called when `pubkey.GetID()` doesn't exists in signatures, maybe assert it's new:
```cpp
auto i = sigdata->signatures.emplace(...);
assert(i.second);
```

achow101 replied:
Done

promag commented on src/script/sign.h:
```diff
@@ -81,4 +91,15 @@ void UpdateInput(CTxIn& input, const SignatureData& data);
+class SignatureExtractorChecker : public BaseSignatureChecker
+{
+private:
+    SignatureData* sigdata;
```
Commit "Make SignatureData able to store signatures and scripts"

Use references instead of pointers?

achow101 replied:
Done

promag commented on src/script/sign.cpp:
```diff
@@ -33,14 +33,60 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
     return true;
 }
 
+static bool GetCScript(const SigningProvider* provider, const SignatureData* sigdata, const CScriptID &scriptid, CScript& script)
```
Commit "Replace CombineSignatures with ProduceSignature"

Looks like these could be refefenreces:
```cpp
static bool GetCScript(const SigningProvider& provider, const SignatureData& sigdata, const CScriptID &scriptid, CScript& script)
```
and remove `provider != nullptr` below.

promag replied:
Otherwise fix space before `scriptid` argument. Same below in the definition.

achow101 replied:
Done

promag commented on src/script/sign.cpp:
```diff
@@ -33,14 +33,60 @@ bool MutableTransactionSignatureCreator::CreateSig(const SigningProvider& provid
+        return true;
+    }
+    // Look for pubkey in all partial sigs
+    const auto& it = sigdata->signatures.find(address);
```
Commit "Replace CombineSignatures with ProduceSignature"

As pointed by @MarcoFalke, don't use references to iterators. Check throughout too.

achow101 replied:
Done.

promag reviewed (commented):
Review comments for:
* fdcbcb380 - Inline Sign1 and SignN
* a86d0679d - Make SignatureData able to store signatures and scripts
* c5d47a58f - Replace CombineSignatures with ProduceSignature
* 95f5a38dd - Remove CombineSignatures and replace tests

fdcbcb380 is kind of unrelated here no?

achow101 commented:
@promag Just so you know, those commits are part of #13425

promag commented:
Ops @achow101 do you want me to comment there?

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+                    }
+
+                    // Read in the signature from value
+                    uint64_t value_len = ReadCompactSize(s);
```
I think this can be written as just `std::vector<unsigned char> sig; s >> sig`.

achow101 replied:
Done

achow101 commented:
No need to comment twice. I will update that PR and then rebase this onto that.

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+
+            // Do stuff based on type
+            switch(type) {
+                case PSBT_IN_NON_WITNESS_UTXO:
```
For this one and the next, check whether you don't already have the other UTXO type?

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+                        throw std::ios_base::failure("Duplicate Key, key for unknown value already provided");
+                    }
+                    // Read in the value
+                    uint64_t value_len = ReadCompactSize(s);
```
Here you can again just use `std::vector<unsigned char> val_bytes; s >> val_bytes;`.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+                    s >> witness_script;
+                    break;
+                }
+                case PSBT_OUT_BIP32_DERIVATION:
```
Can you abstract out the serialization/deserialization code for derivation paths and scripts into separate functions? Otherwise it is needlessly duplicated across input and outputs.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+    std::vector<PartiallySignedInput> inputs;
+    std::vector<PSBTOutput> outputs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    uint64_t num_ins = 0;
```
Do you still need `num_ins` and `use_in_index`?

achow101 replied:
Removed

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+
+        // Read input data
+        unsigned int i = 0;
+        while (!s.empty() && i < tx.vin.size()) {
```
Why the `!s.empty()` check here? If we reach EOF in the stream an error should be raised, not ignore it.

achow101 replied:
If we reach EOF in the stream, this loop will exit and the if block below will throw the error as the number of inputs/outputs will differ from the number of inputs/outputs in the unsigned tx.

sipa replied:
Ah, I see. That will probably give a more useful error.

sipa reviewed (commented):
Just comments on the serialization commit.

achow101 commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+};
+
+/** A structure for PSBTs which contains per output information */
+struct PSBTOutput
```
Done

achow101 commented on src/script/sign.h:
```diff
@@ -73,6 +76,580 @@ struct SignatureData {
+static const uint8_t PSBT_SEPARATOR = 0x00;
+
+template<typename Stream, typename... X>
+void SerializeToVector(Stream& s, const X&... args)
```
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+
+
+// Note: These constants are in reverse byte order because serialization uses LSB
+static constexpr uint32_t PSBT_MAGIC_BYTES = 0x74627370;
```
You could also write it as `uint8_t PSBT_MAGIC_BYTES[4] = "PSBT";`. Byte arrays can be serialized directly now.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+
+// The separator is 0x00. Reading this in means that the unserializer can interpret it
+// as a 0 length key. which indicates that this is the separator. The separator has no value.
+static const uint8_t PSBT_SEPARATOR = 0x00;
```
`constexpr`

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+}
+
+// Deserialize HD keypaths into a map
+static void DeserializeHDKeypaths(const std::vector<unsigned char>& key, std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths, Stream& s)
```
Add `template <typename Stream>` before.

achow101 replied:
Fixed

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+}
+
+// Serialize HD keypaths to a stream from a map
+static void SerializeHDKeypaths(std::vector<uint32_t>>& hd_keypaths, Stream& s)
```
Add `template<typename Stream>` before.

achow101 replied:
Fixed

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,565 @@ struct SignatureData {
+static void SerializeHDKeypaths(std::vector<uint32_t>>& hd_keypaths, Stream& s)
+{
+    for (auto keypath_pair : hd_keypaths) {
+        SerializeToVector(s, PSBT_IN_BIP32_DERIVATION, MakeSpan(keypath_pair.first));
```
Make the field type a function argument; it differs between input and output types.

achow101 replied:
Fixed

sipa commented on src/uint256.h:
```diff
@@ -91,6 +91,15 @@ class base_blob
                ((uint64_t)ptr[7]) << 56;
     }
 
+    uint32_t GetUint32LE(int pos) const
```
You can use `ReadBE32` for this, it's more efficient.

If instead this would be BE (`ReadBE32` also exists), you can get rid of `Uint32ToUint8VectorLE` and instead print using `strprintf("%08x", fingerprint)`.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1262,598 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 3)
+        throw std::runtime_error(
+                            "converttopsbt \"hexstring\" ( permitsigdata iswitness )\n"
+                            "\nConverts a network serialized transaction to a PSBT.\n"
```
Perhaps clarify that it's intended to work with `createrawtransaction` and `fundrawtransaction`, and `createpsbt`/`createfundedpsbt` should be used for new applications.

sipa reviewed (commented):
Comments on rawtransaction.cpp RPCs.

sipa commented:
General comment on naming of RPC field names and input arguments, `"base64"` isn't very informative I think. I suggest using `"psbt"` anywhere you have an input or output in base64. There isn't any other encoding used, so there should never be any confusion. Using `"hex"` for fully signed or legacy partially signed transaction makes sense for consistency with the existing RPCs.

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1262,598 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "  \"bip32_derivs\" : [          (array of json objects, optional)\n"
+            "    {\n"
+            "      \"pubkey\" : \"pubkey\",                     (string) The public key this path corresponds to\n"
+            "      \"master_fingerprint\" : \"fingerprint\"     (string) The fingerprint of the master key\n"
```
I think it is clearer to list the fingerprint separately.

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1262,598 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            }
+        }
+
+        for (const PartiallySignedTransaction& psbtx : psbtxs) {
```
Why?

sipa replied:
Because it's not RPC specific. It's cleaner to put the operations on the lowest level they can go.

achow101 committed:
SignPSBTInput wrapper function

The SignPSBTInput function takes a PSBTInput, SignatureData, SigningProvider,
and other data necessary for signing. It fills the SignatureData with data from
the PSBTInput, retrieves the UTXO from the PSBTInput, signs and finalizes the
input if possible, and then extracts the results from the SignatureData and
puts them back into the PSBTInput.

achow101 committed:
Create utility RPCs for PSBT

decodepsbt takes a PSBT and decodes it to JSON

combinepsbt takes multiple PSBTs for the same tx and combines them.

finalizepsbt takes a PSBT and finalizes the inputs. If all inputs
are final, it extracts the network serialized transaction and returns
that instead of a PSBT unless instructed otherwise.

createpsbt is like createrawtransaction but for PSBTs instead of
raw transactions.

convertpsbt takes a network serialized transaction and converts it
into a psbt. The resulting psbt will lose all signature data and
an explicit flag must be set to allow transactions with signature
data to be converted.

achow101 committed:
Create wallet RPCs for PSBT

walletprocesspsbt takes a PSBT format transaction, updates the
PSBT with any inputs related to this wallet, signs, and finalizes
the transaction. There is also an option to not sign and just
update.

walletcreatefundedpsbt creates a PSBT from user provided data
in the same form as createrawtransaction. It also funds the transaction
and takes an options argument in the same form as fundrawtransaction.
The resulting PSBT is blank with no input or output data filled
in.

sipa commented on src/script/sign.cpp:
```diff
@@ -235,6 +235,36 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
     return sigdata.complete;
 }
 
+bool SignPSBTInput(const SigningProvider& provider, const CTxIn& txin, const CMutableTransaction& tx, PSBTInput& input, SignatureData& sigdata, int index, int sighash, bool sign)
```
No need to pass `txin`; it's always equal to `tx.vin[index]` I think.

achow101 replied:
Done

sipa commented on src/script/sign.cpp:
```diff
@@ -235,6 +235,36 @@ bool ProduceSignature(const SigningProvider& provider, const BaseSignatureCreato
+        utxo = input.non_witness_utxo->vout[txin.prevout.n];
+    }
+    // Now find the witness utxo if the non witness doesn't exist
+    else if (!input.witness_utxo.IsNull()) {
```
Nit: else on the same line as `}`.

achow101 replied:
Fixed

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
     return NullUniValue;
 }
 
+bool parse_hd_keypath(std::string keypath_str, std::vector<uint32_t>& keypath)
```
Nit: function naming style

achow101 replied:
Fixed

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+    return true;
+}
+
+void add_keypath_to_map(const CWallet* pwallet, const CKeyID& keyID, std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths)
```
Nit: function naming style

achow101 replied:
Fixed

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+    hd_keypaths.emplace(vchPubKey, keypath);
+}
+
+bool fill_psbt(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const CTransaction* txConst, int sighash_type, bool sign)
```
Nit: function naming style

achow101 replied:
Fixed

nvk commented:
We will start implementing as soon as this gets merged.
Thanks for sorting it out so fast 👍

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // read size of key
```
This section (up to `s >> MakeSpan(key)`) can be written more concisely as:

```
std::vector<unsigned char> key;
s >> key;
if (key.empty()) return;
```

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+                    if (non_witness_utxo) {
+                        throw std::ios_base::failure("Duplicate Key, input non-witness utxo already provided");
+                    }
+                    if (!witness_utxo.IsNull()) {
```
I guess this is now done by the `IsSane()` function, and doesn't need to be repeated here?

achow101 replied:
Removed

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+                        throw std::ios_base::failure("Duplicate Key, input witness utxo already provided");
+                    }
+                    if (non_witness_utxo) {
+                        throw std::ios_base::failure("A non-witness utxo was already provided, cannot provide a witness utxo");
```
Same thing here.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+                    s >> sig;
+
+                    // Add to list
+                    partial_sigs.emplace(pubkey.GetID(), SigPair(pubkey, sig));
```
Can be `std::move(sig)`.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+    inline void Unserialize(Stream& s) {
+        // Read loop
+        while(!s.empty()) {
+            // read size of key
```
The same shorter read routine is possible here.

achow101 replied:
Done

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,562 @@ struct SignatureData {
+        // Read global data
+        while(!s.empty()) {
+            // read size of key
+            uint64_t key_len = ReadCompactSize(s);
```
And again here.

achow101 replied:
Done

sipa reviewed (commented):
Some more nits on serialization code.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    UniValue result(UniValue::VOBJ);
+    CDataStream ssTx(SER_NETWORK, PROTOCOL_VERSION);
+    ssTx << merged_psbt;
+    return EncodeBase64((unsigned char*)ssTx.data(), ssTx.size());
```
Is the cast necessary here?

achow101 replied:
Yes. `EncodeBase64` takes an `unsigned char*` but `ssTx.data()` is a `char *`.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "createpsbt and walletcreatefundedpsbt should be used for new applications.\n"
+                            "\nArguments:\n"
+                            "1. \"hexstring\"              (string, required) The hex string of a raw transaction\n"
+                            "2. permitsigdata           (boolean, optional) Whether the transaction hex contains any data in the scriptSigs or scriptWitnesses.\n"
```
This explanation is confusing. It's an input argument, right? How about "If true, the RPC will ignore any signatures present in the input. When this argument is false, it will fail if signatures are present."

achow101 replied:
Changed

achow101 commented on src/core_read.cpp:
```diff
@@ -160,6 +161,23 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
+            error = "extra data after PSBT";
+            return false;
+        }
+    } catch (const std::exception& e) {
```
This is checked for above with `if (!ssData.empty()) {`

sipa replied:
I should learn to read.

achow101 commented on src/core_read.cpp:
```diff
@@ -160,6 +161,23 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
 
+bool DecodePSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error)
+{
+    std::vector<unsigned char> txData = DecodeBase64(base64_tx.c_str());
```
`DecodeBase64` takes an `unsigned char*` while base64_tx is a `std::string`. It fails to compile without `c_str()`

sipa replied:
There is also a version of `DecodeBase64` which takes a `const std::string&`.

achow101 replied:
Yes, but not to a `std::vector<unsigned char>`. It decodes to a `std::string`, but I need a `std::vector<unsigned char>`

achow101 commented on src/core_read.cpp:
```diff
@@ -160,6 +161,23 @@ bool DecodeHexBlk(CBlock& block, const std::string& strHexBlk)
 
+bool DecodePSBT(PartiallySignedTransaction& psbt, const std::string& base64_tx, std::string& error)
+{
+    std::vector<unsigned char> txData = DecodeBase64(base64_tx.c_str());
```
Fixed

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
     return result;
 }
 
+static std::string WriteHDKeypath(std::vector<uint32_t>& keypath)
```
Fixed

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            "decodepsbt \"base64string\"\n"
```
Renamed

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "\nCombine multiple partially signed Bitcoin transactions into one transaction.\n"
+            "Implements the Combiner and Finalizer roles.\n"
+            "\nArguments:\n"
+            "1. \"txs\"                   (string) A json array of hex strings of partially signed transactions\n"
```
Fixed

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+    PartiallySignedTransaction merged_psbt(psbtxs[0]); // Copy the first one
+    // Merge them
+    for (const PartiallySignedTransaction& psbtx : psbtxs) {
```
Changed to erase the first element from the vector.

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+        merged_psbt.Merge(psbtx);
+    }
+    if (!merged_psbt.IsSane()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Merged PSBT is not sane");
```
Done

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3607,17 +3529,98 @@ static UniValue fundrawtransaction(const JSONRPCRequest& request)
     result.pushKV("hex", EncodeHexTx(tx));
-    result.pushKV("changepos", changePosition);
-    result.pushKV("fee", ValueFromAmount(nFeeOut));
+    result.pushKV("fee", ValueFromAmount(fee));
```
Why swap these?

achow101 replied:
To match the order in the help text.

sipa commented on src/script/sign.h:
```diff
@@ -73,6 +76,538 @@ struct SignatureData {
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
```
Is this were a `std::unique_ptr<CMutableTransaction>` (or `boost::optional<CMutableTransaction>`), you could avoid the need for implementing `IsNull` or converting to `CTransaction` to decide whether it is set or not.

achow101 replied:
Made into a `boost::optional`

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+    // Get all of the previous transactions
+    bool complete = true;
+    for (unsigned int i = 0; i < txConst->vin.size(); ++i) {
+        CTxIn txin = txConst->vin[i];
```
You can use a const reference here to avoid a copy.

achow101 replied:
Done

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        PSBTInput& input = psbtx.inputs.at(i);
+
+        // If we don't know about this input, skip it and let someone else deal with it
+        uint256 txhash = txin.prevout.hash;
```
Same here.

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1263,553 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    // Get all of the previous transactions
+    bool complete = true;
+    for (unsigned int i = 0; i < psbtx.tx.vin.size(); ++i) {
+        CTxIn txin = psbtx.tx.vin[i];
```
Unused variable.

achow101 replied:
Removed

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+            "\nArguments:\n"
+            "1. \"psbt\"                      (string, required) The transaction base64 string\n"
+            "2. \"sign\",                     (string, optional, default=true) Also sign the transaction when updating\n"
+            "3. \"sighashtype\"            (string, optional, default=ALL) The signature hash type to sign with if not specified by the PSBT. Must be one of\n"
```
I think the semantics should be to fail if the specified sighash type (or the default) does not match the sighash type in the PSBT.

achow101 replied:
Done

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        }
+
+        SignatureData sigdata;
+        complete &= SignPSBTInput(*pwallet, psbtx.tx, input, sigdata, i, sighash, sign);
```
If you'd have a `SigningProvider` wrapper that removes access to private keys, you could use that instead of passing down a `sign` variable. The other call site doesn't need a `sign` anyway, as it's using a dummy provider.

achow101 replied:
Done.

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4388,6 +4391,333 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 5)
+        throw std::runtime_error(
+                            "walletcreatefundedpsbt [{\"txid\":\"id\",\"vout\":n},...] [{\"address\":amount},{\"data\":\"hex\"},...] ( locktime ) ( replaceable ) ( options )\n"
```
Thinking more about this, I can't find a reason why someone would not want to also Update the PSBT. Since in this case, the wallet is adding inputs and change, it is exactly in the position to also add keys/scripts/utxos for those. It shouldn't include signing, though.

achow101 replied:
In order to sign the PSBT, you still have to use `walletupdatepsbt` and that will update the PSBT at the same time.

sipa replied:
I meant: why should createfunded not automatically update? It by definition has all the data available, and I don't think there is a reason why you wouldn't run update immediately after anyway.

achow101 replied:
Changed to also update

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+        }
+        keypath_str += "/";
+    }
+    return keypath_str.substr(0, keypath_str.size() - 1); // Drop last '/'
```
Alternatively: move the `keypath_str += "/";` to the front of the loop and initialize with just `"m"`. That way you can avoid adding and then stripping the final slash.

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "          \"hex\" : \"hex\",            (string) The hex\n"
+            "          \"reqSigs\" : n,            (numeric) The required sigs\n"
+            "          \"type\" : \"pubkeyhash\",    (string) The type, eg 'pubkeyhash'\n"
+            "          \"addresses\" : [           (json array of string)\n"
```
I think it's bad practice to keep outputting "addresses" that correspond to a script. It's confusing (they're not really addresses, just multisig pubkeys represented through their corresponding P2PKH address) and also generally useless as it can't introspect into P2SH and P2WSH.

I preference would be to just output "hex", "asm" and "address" (only if a corresponding address is known through `ExtractDestination`).

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "  \"outputs\" : [                 (array of json objects)\n"
+            "    {\n"
+            "      \"redeem_script\" : {       (json object, optional)\n"
+            "          \"script\" : {           (json object)\n"
```
What is this "script" intermediate layer? I don't see that in the implementation (for this and for witness_script). Again, just "hex" and "asm" is sufficient here I think.

achow101 replied:
Removed, that's a mistake

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+            UniValue out(UniValue::VOBJ);
+
+            out.pushKV("value", ValueFromAmount(txout.nValue));
```
For compatibility with `listunspent` you may want to call this `"amount"`.

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+        // Partial sigs
+        if (!input.partial_sigs.empty()) {
+            UniValue partial_sigs(UniValue::VARR);
```
How about modelling this as an object whose keys are hex pubkeys, and signatures are hex values?

So `{"<p1>":"<sig1>",...}` instead of `[{"pubkey":"<pk1>","signature":""<sig1>"},...]`.

achow101 replied:
Done

sipa replied:
Now you have `[{"<p1>":"<sig1>"},{"<p2>":"<sig2>"},...]`. You can get rid of the internal objects too: `{"<p1>":"<sig1>","<p2>":"<sig2>",...}`.

achow101 replied:
Done.

sipa replied:
It seems you'll need to change the type to `UniValue::VOBJ` above. Calling `pushKV` on an array has no effect (and also doesn't assert/throw anything, it seems...).

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+        // Final scriptSig and scriptwitness
+        if (!input.final_script_sig.empty()) {
+            in.pushKV("final_scriptSig", HexStr(input.final_script_sig));
```
Here you may want to have both `hex` and `asm`.

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    }
+
+    PartiallySignedTransaction merged_psbt(psbtxs[0]); // Copy the first one
+    psbtxs.erase(psbtxs.begin());
```
`psbtxs.pop_front()` will do the same. Alternatively, you can write a loop like `for (auto it = std::next(psbtx.begin()); ++it; it !=psbtx.end()) { merged_psbt.Merge(*it); }`.

achow101 replied:
`std::vector` doesn't have a `pop_front()`, so I used your other suggestion for the loop.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,552 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+    // parse hex string from parameter
+    CMutableTransaction tx;
+    bool try_witness = !request.params[1].isNull() ? (request.params[2].isNull() ? true : request.params[2].get_bool()) : true;
```
These boolean expressions and the ones below are pretty hard to follow.

What about first extracting the boolean values of the arguments, and then writing things in function of those?

achow101 replied:
Changed

sipa commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,341 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        keypath.insert(keypath.begin(), ReadLE32(masterKey.key.GetPubKey().GetID().begin()));
+    }
+    // Single pubkeys get the master fingerprint of themselves
+    else {
```
Style nit: else on the same line

achow101 replied:
Fixed

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "        }\n"
+            "        ,...\n"
+            "      ]\n"
+            "      \"sighash\" : \"type\",                  (string, optional) The sighash type recommended to be used\n"
```
Not just recommended.

achow101 replied:
Fixed

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "          \"hex\" : \"hex\",            (string) The hex\n"
+            "          \"type\" : \"pubkeyhash\",    (string) The type, eg 'pubkeyhash'\n"
+            "        }\n"
+            "       \"bip32_derivs\" : [          (array of json objects, optional)\n"
```
This could be modeled as an object with the pubkey as key, and `{"master_fingerprint":"...","path":"..."}` as value.

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "    }\n"
+            "    ,...\n"
+            "  ]\n"
+            "}\n"
```
It would be very useful to add the fee of the transaction here, if known. Make sure there is a consistency check that non-witness utxo's txid matches the vin.prevout.hash.

achow101 replied:
Done. The consistency check occurs at deserialization time.

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1262,598 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            "combinepsbt [\"base64string\",...]\n"
+            "\nCombine multiple partially signed Bitcoin transactions into one transaction.\n"
```
Well perhaps don't say it implements a Finalizer unless it actually implements that :)

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+            "\nResult:\n"
+            "{\n"
+            "  \"psbt\" : \"value\",          (string) The base64-encoded partially signed transaction\n"
```
That doesn't look like the output it's actually producing.

achow101 replied:
Fixed here and elsewhere

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "4. replaceable               (boolean, optional, default=false) Marks this transaction as BIP125 replaceable.\n"
+                            "                             Allows this transaction to be replaced by a transaction with higher fees. If provided, it is an error if explicit sequence numbers are incompatible.\n"
+                            "\nResult:\n"
+                            "  \"psbt\": \"value\",        (string)  The resulting raw transaction (base64-encoded string)\n"
```
The output looks like just a string, not an object.

achow101 replied:
Fixed here and elsewhere

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1262,6 +1263,541 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "                              If iswitness is not present, heuristic tests will be used in decoding. Only has an effect if\n"
+                            "                              permitsigdata is true\n"
+                            "\nResult:\n"
+                            "  \"psbt\": \"value\",        (string)  The resulting raw transaction (base64-encoded string)\n"
```
Same here.

sipa commented on src/script/sign.h:
```diff
@@ -32,6 +32,17 @@ class SigningProvider
+class PublicOnlySigningProvider : public SigningProvider
+{
+private:
+    const SigningProvider* provider;
```
Style nit: `m_provider`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
```
`> 3`

achow101 replied:
Fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+            {std::string("SINGLE"), int(SIGHASH_SINGLE)},
+            {std::string("SINGLE|ANYONECANPAY"), int(SIGHASH_SINGLE|SIGHASH_ANYONECANPAY)},
+        };
+        std::string strHashType = request.params[1].get_str();
```
`[2]`?

promag replied:
Nit, could move this to a `ParseSigHashType()`? Like `ParseOutputType` and `ParseConfirmTarget`.

achow101 replied:
Moved to a new helper function.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+    // Fill transaction with our data and also sign
+    bool sign = request.params[1].isNull() ? true : request.params[1].get_bool();
+    bool fComplete = FillPSBT(pwallet, psbtx, &txConst, nHashType, sign);
```
nit, just `complete`.

achow101 replied:
Done

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+            + HelpExampleCli("walletprocesspsbt", "\"psbt\"")
+        );
+
+    LOCK2(cs_main, pwallet->cs_wallet);
```
Why lock `cs_main`? Locking could be in `FillPSBT`?

achow101 replied:
Removed `cs_main`, moved to `FillPSBT`.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+            "\nArguments:\n"
+            "1. \"psbt\"                      (string, required) The transaction base64 string\n"
+            "2. \"sign\",                     (string, optional, default=true) Also sign the transaction when updating\n"
```
boolean? Drop `\"` in `sign`.

achow101 replied:
Fixed

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 4)
+        throw std::runtime_error(
+            "walletprocesspsbt \"psbt\" ( sign sighashtype )\n"
```
`"sighashtype"`?

achow101 replied:
Fixed

promag commented on src/rpc/client.cpp:
```diff
@@ -109,6 +109,11 @@ static const CRPCConvertParam vRPCConvertParams[] =
+    { "walletcreatefundedpsbt", 2, "locktime" },
+    { "walletcreatefundedpsbt", 3, "replaceable" },
+    { "walletcreatefundedpsbt", 4, "options" },
     { "createpsbt", 0, "inputs" },
```
Missing `{ "walletprocesspsbt", 1, "sign" }`?

achow101 replied:
Added

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4391,6 +4391,339 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+        // If we don't know about this input, skip it and let someone else deal with it
+        const uint256& txhash = txin.prevout.hash;
+        if (pwallet->mapWallet.count(txhash)) {
```
Use find to avoid 2nd lookup below.

achow101 replied:
Done

promag reviewed (commented):
Comments regarding "Create wallet RPCs for PSBT" (9e4077a)

No `cs_wallet` lock in `walletcreatefundedpsbt`?

Instead of `walletcreatefundedpsbt`, could make a `rawtransactiontopsbt` or alike?

araspitzu commented on src/script/sign.cpp:
```diff
@@ -434,3 +466,160 @@ bool IsSolvable(const SigningProvider& provider, const CScript& script)
+    return redeem_script.empty() && witness_script.empty() && hd_keypaths.empty() && unknown.empty();
+}
+
+void PSBTOutput::Merge(const PSBTOutput& output)
```
Perhaps i'm missing something but shouldn't we merge the unknowns too? (as done in PSBTInput::Merge)

achow101 replied:
Good catch, fixed

instagibbs commented:
Needs rebase now that https://github.com/bitcoin/bitcoin/pull/13425 is merged.

achow101 commented:
Rebased onto master after #13425 was merged. Will address comments soon.

achow101 commented:
> Instead of `walletcreatefundedpsbt`, could make a `rawtransactiontopsbt` or alike?

There already is a `converttopsbt` method. I do not think that we should make using PSBT require using raw transactions first. It should be self contained as I hope that we will move from raw transactions to PSBT entirely.

araspitzu commented on src/script/sign.h:
```diff
@@ -62,24 +77,579 @@ typedef std::pair<CPubKey, std::vector<unsigned char>> SigPair;
+
+// Serialize HD keypaths to a stream from a map
+template<typename Stream>
+void SerializeHDKeypaths(Stream& s, const std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths, uint8_t type)
```
I'm having some trouble matching this code to the spec, with the HD keypaths shouldn't we prefix the "value" (as in the key/value entry) with the fingerprint of the public key? I don't see where it's done here (and neither in 'DeserializeHDKeypaths').

sipa commented on test/functional/rpc_psbt.py:
```diff
@@ -0,0 +1,159 @@
+        pubkey2 = self.nodes[2].getaddressinfo(self.nodes[2].getnewaddress())['pubkey']
+        p2sh = self.nodes[1].addmultisigaddress(2, [pubkey0, pubkey1, pubkey2], "", "legacy")['address']
+        p2wsh = self.nodes[1].addmultisigaddress(2, [pubkey0, pubkey1, pubkey2], "", "bech32")['address']
+        p2wpkh = self.nodes[1].getnewaddress("", "legacy")
```
That doesn't look very witnessy.

achow101 replied:
Fixed. Also added tests for p2pkh and p2sh nested things.

sipa commented:
utACK 14339b0214f2976cae772ae836322d285138c353. Going to play around with it more before I give a tACK.

promag commented:
utACK 14339b0. I have a couple of comments but are all nits here and there.

> There already is a `converttopsbt` method. I do not think that we should make using PSBT require using raw transactions first. It should be self contained as I hope that we will move from raw transactions to PSBT entirely.

Ok got it.

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "Implements the Finalizer and Extractor roles.\n"
+            "\nArguments:\n"
+            "1. \"psbt\"                 (string) A base64 string of a PSBT\n"
+            "2. \"extract\"              (boolean, optional, default=true) If true and the transaction is complete, \n"
```
I'm not sure that this argument is needed. Who is going to use this?

achow101 replied:
It is there so that each role can be done separately instead of at the same time. It's useful for testing.

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    UniValue result(UniValue::VOBJ);
+    CDataStream ssTx(SER_NETWORK, PROTOCOL_VERSION);
+    bool extract = request.params[1].isNull() || (!request.params[1].isNull() && request.params[1].get_bool());
+    if (complete && extract) {
```
help didn't make it clear it will only return one or the other. Might explain why someone would want to turn "extract" off

achow101 replied:
I think the help text is fairly clear that it only has an effect if the transaction is complete. Suggestions for better help text?

instagibbs replied:
something like "extract and return the complete transaction in normal network serialization instead of the PSBT." in addition to my other suggestion in the return values section.

achow101 replied:
Done

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "                             accepted as second parameter.\n"
+                            "   ]\n"
+                            "3. locktime                  (numeric, optional, default=0) Raw locktime. Non-0 value also locktime-activates inputs\n"
+                            "4. replaceable               (boolean, optional, default=false) Marks this transaction as BIP125 replaceable.\n"
```
mu-nit: BIP125-replaceable

Perhaps the default should be true if the sequence values are compatible? It'd be messy to break API later.

achow101 replied:
A lot of this is shared with `createrawtransaction` so changing this will also also cause `createrawtransaction`'s API to change. Having it be separate though means that there will be a lot of code duplication.

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "createpsbt and walletcreatefundedpsbt should be used for new applications.\n"
+                            "\nArguments:\n"
+                            "1. \"hexstring\"              (string, required) The hex string of a raw transaction\n"
+                            "2. permitsigdata           (boolean, optional) If true, any signatures in the input will be discarded.\n"
```
s/will be discarded/will be discarded but processing will continue/

what's the default value?

achow101 replied:
Added default value

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "1. \"hexstring\"              (string, required) The hex string of a raw transaction\n"
+                            "2. permitsigdata           (boolean, optional) If true, any signatures in the input will be discarded.\n"
+                            "                              If false, RPC will fail if any signatures are present.\n"
+                            "3. iswitness               (boolean, optional) Whether the transaction hex is a serialized witness transaction.\n"
```
default value?

achow101 replied:
There really isn't a default value for `iswitness`. Without it being specified, both witness and non-witness are tried. If it is specified, then if it is true, only witness is tried, and false only non-witness is tried.

instagibbs replied:
ok, then please say that in the help

achow101 replied:
Done

instagibbs commented on src/wallet/rpcwallet.h:
```diff
@@ -28,4 +28,5 @@ bool EnsureWalletIsAvailable(CWallet *, bool avoidException);
 
 UniValue getaddressinfo(const JSONRPCRequest& request);
 UniValue signrawtransactionwithwallet(const JSONRPCRequest& request);
+bool FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& psbtx, const CTransaction* txConst, int sighash_type = 1, bool sign = true);
```
why not pass by reference if it's going to be there?

achow101 replied:
Which?

instagibbs replied:
was referring to pwallet, which you dispute as an improvement

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4390,6 +4390,324 @@ UniValue sethdseed(const JSONRPCRequest& request)
+
+            "\nArguments:\n"
+            "1. \"psbt\"                      (string, required) The transaction base64 string\n"
+            "2. sign,                         (boolean, optional, default=true) Also sign the transaction when updating\n"
```
comma after sign?

achow101 replied:
Fixed

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4390,6 +4390,324 @@ UniValue sethdseed(const JSONRPCRequest& request)
+    return true;
+}
+
+void AddKeypathToMap(const CWallet* pwallet, const CKeyID& keyID, std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths)
```
change this to passing wallet by reference as well?

achow101 replied:
Why? It comes as a pointer, why not just pass the pointer around?

instagibbs replied:
Just not crazy about passing around pointers more than we have to.

instagibbs commented:
Left some comments, looking good overall.

I'd also like the RPC help to be a bit more descriptive of what the calls are doing, in plain english, to avoid users having to be BIP174 experts.

laanwj commented:
might want to fix this warning:
```
/.../bitcoin/src/wallet/rpcwallet.h:14:1: warning: class 'PartiallySignedTransaction' was previously declared as a struct [-Wmismatched-tags]
class PartiallySignedTransaction;
^
/.../bitcoin/src/script/sign.h:486:8: note: previous use is here
struct PartiallySignedTransaction
       ^
/.../bitcoin/src/wallet/rpcwallet.h:14:1: note: did you mean struct here?
class PartiallySignedTransaction;
^~~~~
struct
1 warning generated.
```

instagibbs commented on src/script/sign.h:
```diff
@@ -73,6 +77,542 @@ struct SignatureData {
+
+// Deserialize HD keypaths into a map
+template<typename Stream>
+void DeserializeHDKeypaths(Stream& s, const std::vector<unsigned char>& key, std::map<CPubKey, std::vector<uint32_t>>& hd_keypaths)
```
I deleted this comment a while back, github reviews thought it was too good I guess

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+            "    }\n"
+            "    ,...\n"
+            "  ]\n"
+            "  \"fee\" : fee                      (numeric, optional) The transaction fee paid if all UTXOs are present\n"
```
That's not what it means. The fee can be reported whenever the UTXO information for each transaction input is present in the psbt (this information is generally added by Updater roles).

It's clearly confusing though, do you have a better way of formulating this?

instagibbs replied:
something like: "The transaction fee paid if all UTXO slots have been filled in the PSBT"?

achow101 replied:
Done

instagibbs commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1253,6 +1244,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+
+            "\nResult:\n"
+            "{\n"
+            "  \"psbt\" : \"value\",          (string) The base64-encoded partially signed transaction\n"
```
`The base64-encoded partially signed transaction if network encoded transaction not extracted`

achow101 replied:
Done (kind of)

sipa commented:
We should probably have a document (in doc/ or so) that explain the most common workflows (hardware wallet, coinjoin, multisig) with the relevant RPCs.

achow101 commented on src/script/sign.h:
```diff
@@ -73,6 +77,542 @@ struct SignatureData {
+static constexpr uint8_t PSBT_OUT_BIP32_DERIVATION = 0x02;
+
+// The separator is 0x00. Reading this in means that the unserializer can interpret it
+// as a 0 length key. which indicates that this is the separator. The separator has no value.
```
Fixed

achow101 commented on src/script/sign.h:
```diff
@@ -73,6 +77,542 @@ struct SignatureData {
+    }
+}
+
+/** A structure for PSBTs which contain per input information */
```
Fixed

achow101 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,546 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
     return result;
 }
 
+static std::string WriteHDKeypath(std::vector<uint32_t>& keypath)
```
Maybe? It doesn't seem useful outside of this context specifically since keypaths are stored as strings in the wallet currently.

achow101 commented:
> might want to fix this warning

I don't see this warning locally, but fixed anyways.

laanwj commented:
> I don't see this warning locally, but fixed anyways.

Heh, building with clang-latest-master is sometimes useful.

> We should probably have a document (in doc/ or so) that  explain the most common workflows (hardware wallet, coinjoin, multisig)  with the relevant RPCs.

Good idea, though probably not in this PR if we want it in before the feature freeze.

Sjors reviewed (commented):
Lightly tested, great stuff! Though multisig remains a bit tedious and confusing.

>  explain the most common workflows (hardware wallet, coinjoin, multisig) with the relevant RPCs

Adding test for those would be useful, and could catch if we forgot any useful params in the RPC.

I'm missing a test case for `bip32_derivs`.

_Bit of a privacy issue_: let's say person A and B are about to spend from a multisig address, sending everything to person A. Person A gives their address to person B, their wallet wallet creates a  PSBT, but doesn't sign it. Wallet A then calls `walletprocesspsbt` which signs it and  _spontaneously adds the master_fingerprint and bip32 path_. Same issue with `walletcreatefundedpsbt`.

Adding `bip32_derivs` should probably be opt-in.

At the risk over expanding the scope too much, it would be quite useful if `inputs` of `createpsbt` can add add a `redeemScript` argument, as a hint for multisig addresses the wallet might not be aware of. Although see below...

Somewhat orthogonal bug(?): `walletcreatefundedpsbt` is unable to construct a transaction if it can't find the relevant UTXO. E.g. if you call `addmultisig` for a 2 of 2 where you only control one key, and then fund it from an external wallet. Although `getaddressinfo` will be aware of the mutlisig, `getreceivedbyaddress` will claim it's not part of the wallet. You have to import the address and rescan for this to work. It seems to me that the wallet really should care about these transactions even if it doesn't list them.

`walletprocesspsbt` won't sign a transaction if it doesn't know about the multisig address (i.e. addmultisig was never called). This is the case even if the PSBT has sufficient information, since it could match the witness_script against its own pub keys. Might be a nice improvement for later.

 `walletprocesspsbt` also won't sign a transaction if addmultisig was called but it can't find the UTXO (see above for why). However in this case it could in principle blind sign it, or at least throw a more clear error message than `"complete": false`.

These rescans are rather painful, so - if it's not too late - adding a blockheight hint to BIP-174 might be really helpful. I can suggest that on the mailing list if others think it's useful.

It would be nice if all RPC calls which can produce a complete transaction have a flag that lets you produce the hex immediately, rather than going through the `finalizepsbt` dance.

achow101 commented:
> Adding test for those would be useful, and could catch if we forgot any useful params in the RPC.

There already are test cases for multisig. I have added one for coinjoin where inputs come from multiple nodes.

> I'm missing a test case for `bip32_derivs`.

This is done in the updater tests which is a unit test (so we can add specific prevtxs).

> Adding `bip32_derivs` should probably be opt-in.

Done

> At the risk over expanding the scope too much, it would be quite useful if inputs of createpsbt can add add a redeemScript argument, as a hint for multisig addresses the wallet might not be aware of. Although see below...

While that could be useful, `createpsbt` does not do any updating, and I do not think it would be good to include any other data in an input before a UTXO is included for the input.

> Somewhat orthogonal bug(?): walletcreatefundedpsbt is unable to construct a transaction if it can't find the relevant UTXO. E.g. if you call addmultisig for a 2 of 2 where you only control one key, and then fund it from an external wallet. Although getaddressinfo will be aware of the mutlisig, getreceivedbyaddress will claim it's not part of the wallet. You have to import the address and rescan for this to work. It seems to me that the wallet really should care about these transactions even if it doesn't list them.

This is more of a problem with the wallet structure in general. I do not think that PSBT should really be doing anything about this.

> These rescans are rather painful, so - if it's not too late - adding a blockheight hint to BIP-174 might be really helpful. I can suggest that on the mailing list if others think it's useful.

In general, I don't think this is really useful. If you already know where to start rescanning for UTXOs, why can't you just find the UTXO and add it yourself? There isn't really any need to tell someone else to do it.

> It would be nice if all RPC calls which can produce a complete transaction have a flag that lets you produce the hex immediately, rather than going through the finalizepsbt dance.

This was considered, but I think it is easier to follow and understand if the roles are more separated rather than integrated into one command. Instead of having many commands that can produce variable output, we will only have one command that does. The rest will have fixed, known outputs.

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2016 The Bitcoin Core developers
```
Fixed

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -0,0 +1,171 @@
+    def run_test(self):
+
+        # Get some money and activate segwit
+        self.nodes[0].generate(500)
```
Done

achow101 commented on test/functional/rpc_psbt.py:
```diff
@@ -0,0 +1,171 @@
+        new_psbt = self.nodes[0].converttopsbt(rawtx['hex'])
+        self.nodes[0].decodepsbt(new_psbt)
+
+        # BIP 174 Test Vectors
```
Done

sipa commented:
> Bit of a privacy issue: let's say person A and B are about to spend from a multisig address, sending everything to person A. Person A gives their address to person B, their wallet wallet creates a PSBT, but doesn't sign it. Wallet A then calls walletprocesspsbt which signs it and spontaneously adds the master_fingerprint and bip32 path. Same issue with walletcreatefundedpsbt.

I think it's generally hard to maintain much privacy in PSBT. You should assume that those you share a PSBT with will learn more about you than the general public will, even if just by observing which parties add data at what time.

> Somewhat orthogonal bug(?): walletcreatefundedpsbt is unable to construct a transaction if it can't find the relevant UTXO.

That's inevitable. If you don't know what is being spent, you don't know what data to add the the PSBT.

>  walletprocesspsbt won't sign a transaction if it doesn't know about the multisig address (i.e. addmultisig was never called). This is the case even if the PSBT has sufficient information, since it could match the witness_script against its own pub keys. Might be a nice improvement for later.

That sounds like a bug to me. If the PSBT actually has all the information available (UTXO, redeemscript, witnessscript), walletprocesspsbt should sign with all the keys it has.

> However in this case it could in principle blind sign it, or at least throw a more clear error message than "complete": false.

> These rescans are rather painful, so - if it's not too late - adding a blockheight hint to BIP-174 might be really helpful. I can suggest that on the mailing list if others think it's useful.

I think that generally no normal operation should involve rescans. PSBT are for formalizing interactions between pieces of wallet software, and wallet software should already have ways to know about its own UTXOs (which may in exceptional situations mean rescan, but generally shouldn't).

There can also be more low-level separate RPCs that permit simulating some of the functionality without having strict wallets participate. For example:
* `utxoupdatepsbt` (node RPC) which adds UTXO data using chainstate/mempool to a PSBT rather than from a wallet. 
* `descriptorupdatepsbt` (utility RPC) Which is provided an output descriptor (see https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82) and adds derivation paths and scripts.
* `analyzepsbt` (utility RPC, but aware of signing logic) which could report more data on PSBT than just decoding it (it could include fee, estimated vsize after signing, estimated feerate, what stage each input is towards finalizing and/or what's missing to progress further).
* `joinpsbts` (utility RPC) which would combine the inputs and outputs from multiple PSBTs for possible different transactions (effectively constructing a coinjoin), maintaining all utxo/script/derivation data for existing inputs.

achow101 commented:
> `walletprocesspsbt` won't sign a transaction if it doesn't know about the multisig address (i.e. addmultisig was never called). This is the case even if the PSBT has sufficient information, since it could match the witness_script against its own pub keys. Might be a nice improvement for later.

This shouldn't be the case as it is explicitly tested for in the test cases. A multisig address is used with keys from 3 nodes, but the multisig address is only added to one of them. One of the other nodes (without having called `addmultisigaddress`) is still able to sign the transaction and produce a complete, sendable transaction.

> There can also be more low-level separate RPCs that permit simulating some of the functionality without having strict wallets participate

Certainly, but I think such functionality should be done in a separate PR to avoid adding more things to review in this PR which is already somewhat large.

Sjors commented:
In this test case node1 created the multisig and is the first to sign it. In the scenario I tested, the multisig was created by an independent wallet, so the participating wallets only have the matching public and private keys, but don't know there's a multisig involved.

sipa commented:
> In this test case node1 created the multisig and is the first to sign it. In the scenario I tested, the multisig was created by an independent wallet, so the participating wallets only have the matching public and private keys, but don't know there's a multisig involved.

So run walletprocesspsbt on the PSBT with the wallet that does have the information about the multisig construction? It should add it to the file, and permit the signers to use that information.

If I'm misunderstanding the scenario, can you give a script to reproduce the issue? If you're right, this may be an issue.

Sjors commented:
@sipa I described the behavior in more detail in a functional test:
https://github.com/achow101/bitcoin/compare/psbt...Sjors:2018/07/psbt-tests

sipa commented:
@Sjors So what is happening is that none of the wallets involved in your tests treat the address as "mine", and as a result none know about the UTXO being spent, as a result none can add UTXO information to the PSBT, and as a result none can add script information (even those that know about the multisig, because they don't know that it is relevant for that input).

The solution is calling `importaddress` with P2WSH address on one of the nodes, then giving the PSBT to that node, then giving it to the node with the multisig information, and then giving it to all signers.

Sjors commented:
I see, at least we have it documented again :-)

nodar-chkuaselidze commented on src/script/sign.h:
```diff
@@ -62,24 +77,579 @@ typedef std::pair<CPubKey, std::vector<unsigned char>> SigPair;
+            throw std::ios_base::failure("Invalid PSBT magic bytes");
+        }
+        unsigned char magic_sep;
+        s >> magic_sep;
```
Does not `magic_sep` need to be `0xff` ?

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+                            "createpsbt and walletcreatefundedpsbt should be used for new applications.\n"
+                            "\nArguments:\n"
+                            "1. \"hexstring\"              (string, required) The hex string of a raw transaction\n"
+                            "2. permitsigdata           (boolean, optional, default=false) If true, any signatures in the input will be discarde and conversion.\n"
```
Typo: discarde

achow101 replied:
Fixed

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    CMutableTransaction tx;
+    bool permitsigdata = request.params[1].isNull() ? false : request.params[1].get_bool();
+    bool iswitness = request.params[2].isNull() ? true : request.params[2].get_bool();
+    bool try_witness = permitsigdata ? iswitness : true;
```
This seems wrong. With the current code, if permitsigdata is false it will try both witness and no_witness (where it should only permit non-witness, I think?). Further, when iswitness is not specified, it will only try witness.

I think it should be something like:

```c++
bool witness_specified = !request.params[2].isNull();
bool iswitness = witness_specified ? request.params[2].get_bool() : false;
bool try_witness = permitsigdata ? (witness_specified ? iswitness : true) : false;
bool try_no_witness = permitsigdata ? (witness_specified ? !iswitness : true) : true;
```

achow101 replied:
Done

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1259,6 +1260,551 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 4)
+        throw std::runtime_error(
+                            "createpsbt [{\"txid\":\"id\",\"vout\":n},...] [{\"address\":amount},{\"data\":\"hex\"},...] ( locktime ) ( replaceable )\n"
+                            "\nCreates a transaction in the Partially Signed Transaction format. Inputs will be added if supplied inputs are not enough\n"
```
No inputs will be added.

achow101 replied:
Fixed

sipa commented on src/rpc/rawtransaction.cpp:
```diff
@@ -1755,8 +1739,9 @@ UniValue converttopsbt(const JSONRPCRequest& request)
                             "3. iswitness               (boolean, optional) Whether the transaction hex is a serialized witness transaction.\n"
-                            "                              If iswitness is not present, heuristic tests will be used in decoding. Only has an effect if\n"
-                            "                              permitsigdata is true\n"
+                            "                              If iswitness is not present, heuristic tests will be used in decoding. If true, only witness deserializaion\n"
```
Nit: RPC documentation is being changed in unrelated wallet RPC commit.

achow101 replied:
Fixed

sipa commented:
utACK 020628e3a4e88e36647eaf92bac4b3552796ac6a

laanwj commented:
utACK 020628e3a4e88e36647eaf92bac4b3552796ac6a
